AppStoreProject/Common/Extension/UIImageView+Extension.swift

```diff
@@ -2,16 +2,87 @@ import UIKit
+    func setImage(from url: String?, to placeHolder: String = "default.png",path : String) {
+        
+        guard let url = url else {
+            self.image = UIImage(named: placeHolder)
             return
         }
+        
         DispatchQueue.global().async { [weak self] in
-            let data = try? Data(contentsOf: URL(string: imageURLString)!)
-            DispatchQueue.main.async {
-                self?.image = data != nil ? UIImage(data: data!) : UIImage(named: "default.png")
+            
+            
+            //memory caching
+            let cachedKey = NSString(string: url).lastPathComponent
+
+            if let cachedImage = ImageCacheManager.shared.object(forKey: cachedKey as NSString) {
+                self?.image = cachedImage
+                return
+            }
+            
+            
+            //경로 정해주기
+            guard let path = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first else { return }
+            
+            
+            
+            
+            //저장하기
+            var filePath = URL(fileURLWithPath: path)
+            filePath.appendPathComponent(path+cachedKey)
+            
+            guard let urlString = URL(string: url) else {
+                return
+            }
+
+            let data = try? Data(contentsOf: urlString)
+          
+            
+            if !FileManager.default.fileExists(atPath: filePath.path) {
+                FileManager.default.createFile(atPath: filePath.path, contents: data, attributes: nil)
+                
+            }
+            
+            
+//            guard let image = ImageCacheManager.shared.object(forKey: cachedKey as NSString) else { return }
+//            guard let path = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first else { return }
+        
+            
+            //disk caching
+            if FileManager.default.fileExists(atPath: filePath.path) {
+                DispatchQueue.main.async {
+                    
+                    guard let imageData = try? Data(contentsOf: filePath) else { return }
+                    
+                    guard let image = UIImage(data: imageData) else {
+                        return
+                    }
+                    
+                    self?.image = image
+                    
+ 
+                }
+                
+            } else if (ImageCacheManager.shared.object(forKey: cachedKey as NSString) != nil) {
+                
+                self?.image = ImageCacheManager.shared.object(forKey: cachedKey as NSString)
+                
+                
+            } else {
+                
+                DispatchQueue.main.async {
+                    self?.image = data != nil ? UIImage(data: data!) : UIImage(named: placeHolder)
+                    
+                }
+                
+ 
             }
         }
     }
 }
+
+
+class ImageCacheManager {
```

Comment:
ImageCacheManager.shared.downloadImage { downloadedImage in
  self?.image = downloadedImage
}